test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller_phases.go

```diff
@@ -0,0 +1,301 @@
+	if err != nil {
+		return err
+	}
+
+	for _, dockerMachine := range dockerMachinesToDelete {
+		log.V(2).Info("Deleting DockerMachine", "dockerMachine", dockerMachine.Name, "namespace", dockerMachine.Namespace)
+		if err := r.deleteMachinePoolMachine(ctx, dockerMachine, *machinePool); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// deleteMachinePoolMachine attempts to delete a DockerMachine and its associated owner Machine if it exists.
+func (r *DockerMachinePoolReconciler) deleteMachinePoolMachine(ctx context.Context, dockerMachine infrav1.DockerMachine, machinePool expv1.MachinePool) error {
+	log := ctrl.LoggerFrom(ctx)
+	machine, err := util.GetOwnerMachine(ctx, r.Client, dockerMachine.ObjectMeta)
+	// Not found doesn't return an error, so we need to check for nil.
+	if err != nil {
+		return errors.Wrapf(err, "error getting owner Machine for DockerMachine %s/%s", dockerMachine.Namespace, dockerMachine.Name)
+	}
+	if machine == nil {
+		// If MachinePool is deleted, the DockerMachine and owner Machine doesn't already exist, then it will never come online.
+		if mpDeleted := isMachinePoolDeleted(ctx, r.Client, &machinePool); mpDeleted {
+			log.Info("DockerMachine is orphaned and MachinePool is deleted, deleting DockerMachine", "dockerMachine", dockerMachine.Name, "namespace", dockerMachine.Namespace)
+			if err := r.Client.Delete(ctx, &dockerMachine); err != nil {
+				return errors.Wrapf(err, "failed to delete orphaned DockerMachine %s/%s", dockerMachine.Namespace, dockerMachine.Name)
+			}
+		} else { // If the MachinePool still exists, then the Machine will be created, so we need to wait for that to happen.
+			return errors.Errorf("DockerMachine %s/%s has no owner Machine, will reattempt deletion once owner Machine is present", dockerMachine.Namespace, dockerMachine.Name)
+		}
+	} else {
+		if err := r.Client.Delete(ctx, machine); err != nil {
+			return errors.Wrapf(err, "failed to delete Machine %s/%s", machine.Namespace, machine.Name)
+		}
+	}
+
+	return nil
+}
+
+// getDockerMachinesToDelete returns a list of DockerMachines representing excess replicas and Docker containers that do not match the infrastructure spec.
+// In the case of excess replicas, DockerMachines with the clusterv1.DeleteMachineAnnotation are prioritized to ensure they are deleted first.
+func getDockerMachinesToDelete(ctx context.Context, dockerMachines []infrav1.DockerMachine, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool) ([]infrav1.DockerMachine, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	dockerMachinesToDelete := []infrav1.DockerMachine{}
+	labelFilters := map[string]string{dockerMachinePoolLabel: dockerMachinePool.Name}
+
+	// Sort priority delete to end of list
+	sort.Slice(dockerMachines, func(i, j int) bool {
+		_, iHasAnnotation := dockerMachines[i].Annotations[clusterv1.DeleteMachineAnnotation]
```

Comment:
Is this part of docker machine pool machine core logic?

Reply:
Yes, as part of the proposal, we want to prioritize deleting Machines with the `clusterv1.DeleteMachineAnnotation` so MachinePools can work with cluster autoscaler. In this case, the DockerMachine controller sets the annotation on itself if its owner Machine has it too.